### PR TITLE
Undocumented APIs can output as structured JSON

### PR DESCRIPTION
**Pull request by jverkoey**
When provided, a tab-separated table of linter-specific results will be written to the file. If a relative path is provided, the path will be considered relative to the current working directory.

Report format

- The first column of each row defines the report type.
- The number of additional columns is dependent on the report type.

The contents of each report type are outlined below (commas represent tabs):

```
<report type>, <columns>
 undocumented, full_path, line_number, symbol, symbol_type
```

This PR removes the undocumented.txt file output in favor of the lint-report output.

The intent of this option is to enable a linting engine to communicate with jazzy. Future linting results should be possible using the "report type" field that precedes each column.

### CODE REVIEW

**Comment by jverkoey** [comment]
This is intended to replace #390.

**Comment by jverkoey** [comment]
What's Jazzy's approach to deprecating previous functionality? I worry that people may be depending on the undocumented.txt file in some manner. I'm happy to leave the code that generates it here if there's a more graceful way to remove that behavior.

**Comment by jverkoey** [comment]
#393 is related to the workflow of:

    jazzy --lint-report=/tmp/report.tsv --skip-documentation

Which allows a linting engine to gather information without generating the site.

**Comment by pcantrell** [comment]
I like this approach. It seems flexible enough to accommodate different uses, it leaves room for future expansion, and the options mean what you’d assume they mean.

One last annoying question on the approach: is it normal to have a `.tsv` with (potentially) multiple sections with different columns? Would that be hard to code against? I wonder if the output should be `.json` instead, e.g.:

    [
      {
        "file": "Source/Foo.swift",
        "line": "310",
        "warning": "undocumented",
        "symbol": "Foo",
        "symbol_type": "class"
      }
    ]

File: lib/jazzy/config.rb [hunk 1]
```diff
@@ -226,6 +226,19 @@ def expand_path(path)
                    'comments.',
       default: false
 
+    config_attr :lint_report,
+      description: ['When provided, a tab-separated table of results will be written',
+                    'to the file. If a relative path is provided, the path will be ',
+                    'considered relative to the current working directory.',
```
Comment (pcantrell) [inline lib/jazzy/config.rb]: This isn’t quite correct. Because you (rightly) use `expand_path`, it's relative to working dir if specified on the command line, relative to the config file's parent if specified in a config file.

Since this is standard for all path-based options and is explained in the online help, no need to explain it here.

**Comment by jverkoey** [comment]
Was debating using json as well - was mostly nervous that json's a bit harder to script against. Tabs are easier to `cut`.

Perhaps I fixated too much on it being in a single file. What if it output to a directory?

    jazzy --lint-report=/tmp/report/ --skip-documentation

Could generate

    /tmp/report/undocumented.tsv
    /tmp/report/lint-warnings.tsv
    ...

File: lib/jazzy/config.rb (hunk 1 above)
Comment (jverkoey) [inline lib/jazzy/config.rb]: Oh! That's good to know.

**Comment by jverkoey** [comment]
Ah, it may be reasonable at some point to allow the file format to be configured. `--lint-report-type=json` or `--lint-report-type=tsv`, etc...

**Comment by pcantrell** [comment]
Not a fan of the multiple files; that doesn't seem simpler. I imagine, for example, a tool that generates an HTML report, or one that annotates the source on Github. Both of those would be easier to write with JSON with standard file & line keys.

I think in this day and age etc etc, people are as likely to be doing the processing in Ruby, JS, Python or even Swift as they are in raw bash.

You could certainly infer the output type from the file extension (as the config file code does), but supporting multiple output types right now seems to me like crossing too many bridges at once.

**Comment by jverkoey** [comment]
Happy to go with json :)

**Comment by jverkoey** [comment]
Updated to write out a json object.

I use `source_directory` to avoid re-running jazzy on a directory multiple times from the linting engine.

**Comment by pcantrell** [comment]
The JSON structure looks good to me. Tried it on Siesta, and I like the output.

All the line numbers are coming out null. On inspection, it looks like SourceKitten itself is not emitting line numbers for undocumented declarations. @jpsim, does that sound plausible?

**Comment by jverkoey** [comment]
Following up on this guy - what can I do to help push it along?

File: lib/jazzy/doc_builder.rb [hunk 2]
```diff
@@ -165,12 +170,25 @@ def self.write_undocumented_file(undocumented, output_dir)
...
+              symbol: token['key.name'],
+              symbol_kind: token['key.kind'],
+              warning: 'undocumented'
+            }
           end
         end
+
+        lint_report = {
+          warnings: warnings,
+          source_directory: options.source_directory
```
Comment (jpsim) [inline lib/jazzy/doc_builder.rb]: why is including the source directory useful?

File: lib/jazzy/doc_builder.rb [hunk 3]
```diff
@@ -113,6 +113,10 @@ def self.build_docs_for_sourcekitten_output(sourcekitten_output, options)
         options.skip_undocumented,
       )
 
+      if options.lint_report
```
Comment (jpsim) [inline lib/jazzy/doc_builder.rb]: why was this moved up?

File: lib/jazzy/config.rb [hunk 4]
```diff
@@ -226,6 +226,17 @@ def expand_path(path)
                    'comments.',
       default: false
 
+    config_attr :lint_report,
+      description: ['Sets the path to which a lint report will be written. If',
+                    'no path is provided, no lint report will be written.',
```
Comment (jpsim) [inline lib/jazzy/config.rb]: writing this file is fairly cheap compared to the rest of the disk IO jazzy does, so why not just default to `undocumented.json` where the current `undocumented.txt` would be?

File: lib/jazzy/doc_builder.rb (hunk 2 above)
Comment (jverkoey) [inline lib/jazzy/doc_builder.rb]: The linting engine I'm using calls a linter once for each file that was changed. I use `source_directory` to avoid re-running jazzy on the same project multiple times.

File: lib/jazzy/config.rb (hunk 4 above)
Comment (jverkoey) [inline lib/jazzy/config.rb]: If the lint report ends up growing in complexity in the future I worry that we'd regret having enabled it by default. Thoughts?

**Comment by jpsim** [comment]
Sorry it's taken me a while to review this @jverkoey, I like where this is going! The format of the undocumented.txt was always gnarly.

I'm wondering though if we could get away with not having a config option for this at all, just always writing `undocumented.json` like we do now for `undocumented.txt`. The user of jazzy can just move the file after its generation, which would alleviate the need for this complexity in jazzy itself and be mostly the same effort as a flag for users.

File: lib/jazzy/doc_builder.rb (hunk 3 above)
Comment (jverkoey) [inline lib/jazzy/doc_builder.rb]: The intent is to have the lint report be a separate phase from the doc generation (see #393).

File: lib/jazzy/doc_builder.rb (hunk 2 above)
Comment (pcantrell) [inline lib/jazzy/doc_builder.rb]: Including it solves another problem too: do you write full paths to favor easy follow-up with the files on disk, or source-relative paths to favor e.g. generating Github links? This approach lets us have our cake and eat it too.

Comment (jpsim) [inline lib/jazzy/doc_builder.rb]: cool :+1:

File: lib/jazzy/doc_builder.rb (hunk 3 above)
Comment (jpsim) [inline lib/jazzy/doc_builder.rb]: fair enough :+1:

File: lib/jazzy/config.rb (hunk 4 above)
Comment (jpsim) [inline lib/jazzy/config.rb]: We'll cross that bridge when (if) we get there?

**Comment by jpsim** [comment]
Yeah, I still think generating the report by default, in the current location, is the way to go for this.

**Comment by jverkoey** [comment]
np - happy to keep shepherding it along.

Do you expect that people would want the lint report enabled in common scenarios? I generally err on the side of conservatism with new features - thus the flag - but I'll make it default to on if that'll be preferred :)

**Comment by jpsim** [comment]
I don't consider this a new feature though, despite it being quite useful, it's really just a different format for the current functionality.

**Comment by jverkoey** [comment]
Removed the flag - now generated in the docs folder by default.

In a related note: should we deprecate the `undocumented.txt` file in this change and remove it in a subsequent release of Jazzy?

**Comment by jpsim** [comment]
Looks like `output_dir` isn't defined now:

```diff
-jam out ♪♫ to your fresh new docs in `docs`
+ROOT/lib/jazzy/doc_builder.rb:160:in `write_lint_report': undefined local variable or method `output_dir' for Jazzy::DocBuilder:Module (NameError)
+	from ROOT/lib/jazzy/doc_builder.rb:116:in `build_docs_for_sourcekitten_output'
+	from ROOT/lib/jazzy/doc_builder.rb:68:in `build'
+	from ROOT/bin/jazzy:15:in `<main>'
```

No, I don't think it's necessary to continue generating `undocumented.txt`, but a changelog entry explaining this change would be nice.

**Comment by jverkoey** [comment]
Updated CHANGELOG and fixed output.

**Comment by jverkoey** [comment]
I am looking into getting this PR landed again. I've attempted to run the spec regeneration steps as outlined in `CONTRIBUTING.md` but encounter similar failures as the ones that Travis outputs:

```
      ✓ docs/docsets/JazzyKit.docset/Contents/Resources/Documents/js/jquery.min.js (2 ms)
      - docs/docsets/JazzyKit.docset/Contents/Resources/Documents/undocumented.txt [FAILED]
      ✓ docs/docsets/JazzyKit.docset/Contents/Resources/docSet.dsidx.csv
      ✓ docs/img/carat.png
      ✓ docs/img/dash.png
      ✓ docs/img/gh.png
      ✓ docs/index.html
      ✓ docs/js/jazzy.js
      ✓ docs/js/jquery.min.js (2 ms)
      - docs/undocumented.txt [FAILED]
      ✓ execution_output.txt
      - should not produce unexpected files [FAILED]
```

I ran `bundle exec rake rebuild_integration_fixtures` and expected it to generate the docs for each test case, at which point I'd evaluate the diffs and check in the results to the specs repo.

Am I missing a step?

**Comment by segiddins** [comment]
`bundle exec rake rebuild_integration_fixtures` will show the failures, but if you run the specs again they should pass

**Comment by jverkoey** [comment]
The final output of `bundle exec rake rebuild_integration_fixtures` is:

```
398 specifications (770 requirements), 18 failures, 0 errors
rake aborted!
Command failed with status (1): [bundle exec bacon spec/integration_spec.rb...]
jazzy/Rakefile:43:in `block in <top (required)>'
Tasks: TOP => rebuild_integration_fixtures
(See full trace by running task with --trace)
```

No changes seem to have been made to the `integration_specs/` directories; is that expected?

**Comment by jverkoey** [comment]
I'm assuming that to run the specs again I run `bundle exec rake spec`.

The output for that, after running the above, still outputs errors regarding missing `undocumented.txt` files.

```
398 specifications (770 requirements), 18 failures, 0 errors
rake aborted!
Command failed with status (1): [bundle exec bacon spec/integration_spec.rb...]
jazzy/Rakefile:34:in `block in <top (required)>'
Tasks: TOP => spec
(See full trace by running task with --trace)
```

Pardon my total lack of knowledge for running Ruby tests; here are the assumptions I'm making that might be leading me down the wrong path:

- The tests run jazzy on the `before` directory for each spec. The output from this is compared again the `after` directory.
- Anything that differs between the before's generated results and the after directory counts as a failure.
- The purpose of `rebuild_integration_fixtures` is to regenerate the output for every "before" spec.
- The output from `rebuild_integration_fixtures` replaces the contents of the `after` directory. As per the CONTRIBUTING.md, the expectation is to then commit and push this output for the purposes of running CI.

I'm not sure where to look to learn how to move forward from here.

**Comment by jverkoey** [comment]
I think I may have found the source of the problem.

I had started running `bundle exec rake rebuild_integration_fixtures` and wasn't sure if the process had hung, so I went into the Rakefile and changed line 43's `puts` to a `sh` so that I could see output as it happened. I'm assuming that if the `sh` command fails in this context it causes the output as shown above and the rake terminates.

I've reverted my change to the Rakefile and will patiently wait the results this time :)

**Comment by jverkoey** [comment]
That was it! Curse my impatience >.<

**Comment by jverkoey** [comment]
Yay! We're green and good to go :)

@jpsim I'll leave it to you to have final say on merging.

**Comment by jpsim** [comment]
Nice work :+1:
